### PR TITLE
External name to parameters mapping

### PR DESCRIPTION
**Pull request by muvaf**
### Description of your changes

First stab at opening up an interface for per-resource configuration and external name function is the first implementation. As described in https://github.com/crossplane-contrib/terrajet/issues/11#issuecomment-919154858 we're going to ask authors to provide a set of fields to omit from top level spec and give a function that will take external name and set the necessary fields in the output HCL before actual parameters are applied.

Fixes https://github.com/crossplane-contrib/terrajet/issues/11

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

Code compiles with provider-tf-aws but I haven't tried provisioning a resource. See https://github.com/crossplane-contrib/provider-tf-aws/pull/19

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by muvaf** [comment]
I investigated ways to avoid getting a type path but it doesn't seem possible. Hence opted for having a `SelfVarPath` in the struct that the users will give so that we can have some signature checking in compilation time since the func is part of that struct now and the path string of `var` will be close to the `var` itself.

File: pkg/terraform/resource/configuration.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+// ConfigureWithNameFn functions configure the base resource fields by using
+// given name value.
+type ConfigureWithNameFn func(base map[string]interface{}, name string)
+
+// ConfigurationOption allows setting optional fields of a Configuration object.
+type ConfigurationOption func(*Configuration)
+
+// WithExternalName allows you to set an ExternalNamer for given Configuration.
+func WithExternalName(e ExternalNamer) ConfigurationOption {
```
Comment (turkenh) [inline pkg/terraform/resource/configuration.go:27]:
```suggestion
func WithExternalNamer(e ExternalNamer) ConfigurationOption {
```

File: pkg/terraform/resource/configuration.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	for _, f := range opts {
+		f(c)
+	}
+	return c
+}
+
+// ExternalNamer contains all information that is necessary for naming operations,
+// such as removal of those fields from spec schema and calling Configure function
+// to fill attributes with information given in external name.
+type ExternalNamer struct {
```
Comment (turkenh) [inline pkg/terraform/resource/configuration.go:57]: `ExternalNamer` sounds like it contains configuration/information to build the external name (using existing fields), whereas, it is mostly for almost the opposite, to parse the external name into spec. Could we find a more appropriate name for this type to prevent possible confusion, e.g. `NamingConfiguration` ?

File: pkg/tfcli/templates/main.tf.json.tpl [hunk 3]
```diff
@@ -20,11 +20,6 @@
         "{{ .Resource.LabelType }}": {
             "{{ .Resource.LabelName }}": {
                 {{ .Resource.Body | printf "%s" }}
-                {{ if .Lifecycle.PreventDestroy -}}
```
Comment (turkenh) [inline pkg/tfcli/templates/main.tf.json.tpl:23]: Are these relevant to this PR? Or just a refactor?

**Review by turkenh — APPROVED** [review]
LGTM! Thanks @muvaf !

File: pkg/terraform/resource/configuration.go (hunk 2 above)
Comment (muvaf) [reply pkg/terraform/resource/configuration.go:61]: Named the type as `ExternalNameConfiguration` but kept the field name in `Configuration` as `ExternalName`

File: pkg/tfcli/templates/main.tf.json.tpl (hunk 3 above)
Comment (muvaf) [reply pkg/tfcli/templates/main.tf.json.tpl:23]: Here we actually do some string manipulation on the printed form of the JSON, see the `,` before `prevent_destroy`. We'd have to unmarshal/marshal from `[]byte` to avoid doing that but now that we get to work with `map[string]interface{}`, we don't need any of this anymore. The additional property added directly to the map which is printed as is.

**Comment by muvaf** [comment]
We synced with @turkenh about the `SelfVarPath` and how we could avoid it in the context of reference configurations. My main concern was that if we use a `map[string]resource.Configuration` in `apis` package,  then it'd mean that the `resource.Configuration` object would be limited to what can be imported in `apis` package which is quite limited. But @turkenh brought up the idea that we can have another struct that includes `resource.Configuration` so that its scope would be limited to only API type operations only, which would eliminate my concern.

The map will be similar to scheme accumulation in `apis` package, we decided to do it as part of https://github.com/crossplane-contrib/terrajet/issues/56
